### ISSUE TITLE
(0.99.14 Linux) Tables disappear and reappears when I switch nodes or when I restart the app

### ISSUE THREAD

**Issue by Salvaju29ro** [giuspen/cherrytree#1205]
Tables appear and disappear when I switch nodes or when I restart the app

![image](https://user-images.githubusercontent.com/10899871/94926340-ec529880-04c0-11eb-8591-d4781cdfb266.png)

When I restart it disappears
If then I restart again it reappears again

![image](https://user-images.githubusercontent.com/10899871/94926367-f70d2d80-04c0-11eb-9143-e63de1fb3941.png)

**Comment by ciranus** [comment]
I cannot reproduce with 0.99.14 in linux/KDE.
To be confirmed.

Which linux distribution ?

**Comment by Salvaju29ro** [comment]
> I cannot reproduce with 0.99.14 in linux/KDE.
> To be confirmed.
> 
> Which linux distribution ?

Operating System: Arch Linux
KDE Plasma Version: 5.19.5
KDE Frameworks Version: 5.74.0
Qt Version: 5.15.1
Kernel Version: 5.8.12-arch1-1
OS Type: 64-bit
Processors: 4 × Intel® Core™ i3-5005U CPU @ 2.00GHz
Memory: 7.7 GiB of RAM
Graphics Processor: Mesa Intel® HD Graphics 5500

**Comment by ciranus** [comment]
Then https://aur.archlinux.org/packages/cherrytree/ . 
Are you the only Arch user observing this issue ? (tbc on arch forum ?).
It looks like a drawing issue because the table data is not lost.

**Comment by Salvaju29ro** [comment]
> Quindi https://aur.archlinux.org/packages/cherrytree/ .
> Sei l'unico utente di Arch a osservare questo problema? (da confermare su arch forum?).
> Sembra un problema di disegno perché i dati della tabella non vengono persi.

Yes, I installed it from aur

I honestly don't know if other Arch users have this problem

They are not lost
It happened to me that the bottom table ended up behind the top one
But that's not a serious problem, because they reappear when the app is restarted

**Comment by ciranus** [comment]
Can you try with a different theme ? (see Preferences).
Or reset the gtk3 default settings ?

**Comment by txe** [comment]
@Salvaju29ro , i saw the issue a few times, though usually switching between nodes puts tables in their places. I'll take a look

**Comment by Salvaju29ro** [comment]
> Puoi provare con un tema diverso? (vedi Preferenze).
> O ripristinare le impostazioni predefinite di gtk3?
> `~/.config/gtk-3.0/`
> 
> O anche con una nuova configurazione CT?
> `~/.config/cherrytree/`

Thanks for the suggest, first of all.

I tried with a new cherrytree configuration and the problem reappeared, BUT as @txe said, switching from one node to another the table reappears
I also tried to change the theme (I changed the theme from Breeze to Default) and the problem returns.
But, as I said, it's not that bad because when you change the node it comes back

**Comment by ciranus** [comment]
I definitively cannot reproduce this at all is my environment (currently ubuntu groovy / KDE).
Strange.

**Comment by Salvaju29ro** [comment]
> I definitively cannot reproduce this at all is my environment (currently ubuntu groovy / KDE).
> Strange.

Honestly, I don't know what the reason is
@txe Do you have arch linux too? Maybe it's a problem that exists only on Arch

**Comment by txe** [comment]
Yeah, I have Arch(manjaro) with XFCE

**Comment by ciranus** [comment]
@Salvaju29ro 
If you create a 100% new instance, create a table (no copy/paste from previous table), can you reproduce this issue ?

**Comment by Salvaju29ro** [comment]
> @Salvaju29ro
> If you create a 100% new instance, create a table (no copy/paste from previous table), can you reproduce this issue ?

Trying with a new session and with new tables they don't seem to disappear
I will try for longer

**Comment by ciranus** [comment]
These tables with problems, did you create them within cherrytree, or from some copy/paste from somewhere outside CT ?

**Comment by Salvaju29ro** [comment]
> These tables with problems, did you create them within cherrytree, or from some copy/paste from somewhere outside CT ?

With Cherrytree, but with the old version written in gtk2

**Comment by ciranus** [comment]
What you can do for investigation, is to export the two nodes with tables (old&new) as **Cherrytree** document > **XML** format, open the xml file in a text editor, and see you don't find differences which could explain the gtk2/gtk3 table difference (if any).
Maybe useless investigation, but who knows....

**Comment by Salvaju29ro** [comment]
> What you can do for investigation, is to export the two nodes with tables (old&new) as **Cherrytree** document > **XML** format, open the xml file in a text editor, and see you don't find differences which could explain the gtk2/gtk3 table difference (if any).
> Maybe useless investigation, but who knows....

Sorry for my incompetence, but when I export an xml file (archive type) it saves it to me in .ctd, is this normal or am I wrong something?

**Comment by ciranus** [comment]
Yes that's normal, but it does not prevent you to open this file with a text editor.

**Comment by Salvaju29ro** [comment]
These are the tables created with the old version (I opened the CTD file with Kate)

```
<?xml version="1.0" encoding="UTF-8"?>
<cherrytree>
  <node name="Serie in Corso" unique_id="1" prog_lang="custom-colors" tags="" readonly="0" custom_icon_id="0" is_bold="0" foreground="" ts_creation="1579528353" ts_lastsave="1601716318">
    <rich_text justification="left"></rich_text>
    <rich_text>

</rich_text>
    <rich_text scale="h2" foreground="#ffffffff0000" weight="heavy">Anime in Corso</rich_text>
    <rich_text>

</rich_text>
    <rich_text justification="left"></rich_text>
    <rich_text>


</rich_text>
    <table char_offset="0" justification="left" col_min="300" col_max="300">
      <row>
        <cell>The Good Doctor</cell>
        <cell>3</cell>
        <cell>20</cell>
      </row>
      <row>
        <cell>Gomorra</cell>
        <cell>4</cell>
        <cell>12</cell>
      </row>
      <row>
        <cell>Better Call Saul</cell>
        <cell>4</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>The 100</cell>
        <cell>6</cell>
        <cell>13</cell>
      </row>
      <row>
        <cell>Stranger Things</cell>
        <cell>3</cell>
        <cell>8</cell>
      </row>
      <row>
        <cell>Narcos: Messico</cell>
        <cell>2</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>Elite</cell>
        <cell>3</cell>
        <cell>8</cell>
      </row>
      <row>
        <cell>Sex Education</cell>
        <cell>2</cell>
        <cell>8</cell>
      </row>
      <row>
        <cell>Love, Victor</cell>
        <cell>1</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>Prison Break</cell>
        <cell>1</cell>
        <cell>22</cell>
      </row>
      <row>
        <cell>Blu Notte</cell>
        <cell>2</cell>
        <cell>1</cell>
      </row>
      <row>
        <cell>Skam</cell>
        <cell>4</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>Superstore</cell>
        <cell>2</cell>
        <cell>7</cell>
      </row>
      <row>
        <cell>Serie</cell>
        <cell>Stagione</cell>
        <cell>Episodio</cell>
      </row>
    </table>
    <table char_offset="19" justification="left" col_min="300" col_max="300">
      <row>
        <cell>Ace of Diamond Act II</cell>
        <cell>3</cell>
        <cell>46</cell>
      </row>
      <row>
        <cell>Haykyuu</cell>
        <cell>4</cell>
        <cell>13</cell>
      </row>
      <row>
        <cell>Major 2nd</cell>
        <cell>1</cell>
        <cell>25</cell>
      </row>
      <row>
        <cell>Boku no Hero Academia</cell>
        <cell>4</cell>
        <cell>25</cell>
      </row>
      <row>
        <cell>Detective Conan</cell>
        <cell></cell>
        <cell>495</cell>
      </row>
      <row>
        <cell>Chihayafuru</cell>
        <cell>3</cell>
        <cell>5</cell>
      </row>
      <row>
        <cell>Oreigaru</cell>
        <cell>3</cell>
        <cell>6</cell>
      </row>
      <row>
        <cell>Serie</cell>
        <cell>Stagione</cell>
        <cell>Episodi</cell>
      </row>
    </table>
  </node>
</cherrytree>
```

These are the tables created with the new version

```
<?xml version="1.0" encoding="UTF-8"?>
<cherrytree>
  <node name="Serie in Corso" unique_id="4" prog_lang="custom-colors" tags="" readonly="0" custom_icon_id="0" is_bold="0" foreground="" ts_creation="1601715990" ts_lastsave="1601716344">
    <rich_text justification="left"></rich_text>
    <rich_text>

</rich_text>
    <rich_text scale="h2" foreground="#ffffffff0000" weight="heavy">Anime in Corso</rich_text>
    <rich_text>

</rich_text>
    <rich_text justification="left"></rich_text>
    <table char_offset="0" justification="left" col_min="300" col_max="300">
      <row>
        <cell>The Good Doctor</cell>
        <cell>3</cell>
        <cell>20</cell>
      </row>
      <row>
        <cell>Gomorra</cell>
        <cell>4</cell>
        <cell>12</cell>
      </row>
      <row>
        <cell>Better Call Saul</cell>
        <cell>4</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>The 100</cell>
        <cell>6</cell>
        <cell>13</cell>
      </row>
      <row>
        <cell>Stranger Things</cell>
        <cell>3</cell>
        <cell>8</cell>
      </row>
      <row>
        <cell>Narcos: Messico</cell>
        <cell>2</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>Elite</cell>
        <cell>3</cell>
        <cell>8</cell>
      </row>
      <row>
        <cell>Sex Education</cell>
        <cell>2</cell>
        <cell>8</cell>
      </row>
      <row>
        <cell>Love, Victor</cell>
        <cell>1</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>Prison Break</cell>
        <cell>1</cell>
        <cell>22</cell>
      </row>
      <row>
        <cell>Blu Notte</cell>
        <cell>2</cell>
        <cell>1</cell>
      </row>
      <row>
        <cell>Skam</cell>
        <cell>4</cell>
        <cell>10</cell>
      </row>
      <row>
        <cell>Superstore</cell>
        <cell>2</cell>
        <cell>7</cell>
      </row>
      <row>
        <cell>Serie	</cell>
        <cell>Stagione</cell>
        <cell>Episodio</cell>
      </row>
    </table>
    <table char_offset="19" justification="left" col_min="300" col_max="300">
      <row>
        <cell>Ace of Diamond Act II</cell>
        <cell>3</cell>
        <cell>46</cell>
      </row>
      <row>
        <cell>Haykyuu</cell>
        <cell>4</cell>
        <cell>13</cell>
      </row>
      <row>
        <cell>Major 2nd</cell>
        <cell>1</cell>
        <cell>25</cell>
      </row>
      <row>
        <cell>Boku no Hero Academia</cell>
        <cell>4</cell>
        <cell>25</cell>
      </row>
      <row>
        <cell>Detective Conan</cell>
        <cell></cell>
        <cell>495</cell>
      </row>
      <row>
        <cell>Chihayafuru</cell>
        <cell>3</cell>
        <cell>5</cell>
      </row>
      <row>
        <cell>Oreigaru</cell>
        <cell>3</cell>
        <cell>6</cell>
      </row>
      <row>
        <cell>Serie</cell>
        <cell>Stagione</cell>
        <cell>Episodi</cell>
      </row>
    </table>
  </node>
</cherrytree>
```

The old version has 132 lines, while the new one has 128

**Comment by ciranus** [comment]
I have imported your "old" table (saved as old-table.ctd), and don't have any problem with it.
The tables are finally maybe not the problem.

**Comment by Salvaju29ro** [comment]
Two in one nodes
Now I have tried to delete the node with the tables created with the old version and replaced it with the node with the tables created with the new version. Just to try.

**Comment by ciranus** [comment]
Only 2 tables ??? Then if creating new table does not reproduce this issue, you have an easy solution.
A problem impossible to reproduce cannot be solved.

**Comment by Salvaju29ro** [comment]
I tried to create a new node with two tables
When the program starts, the tables overlap:

![image](https://user-images.githubusercontent.com/10899871/94990248-d0abc880-057a-11eb-902c-c20d3d2a77a4.png)

When I change node and return to the one with tables, it returns to normal:

![image](https://user-images.githubusercontent.com/10899871/94990255-e3260200-057a-11eb-8ce3-eb2999fecf45.png)

If it's my problem and it's not a bug, it's not a problem, because I can live with it without problems
It's not that bad, because the tables don't disappear permanently

**Comment by ciranus** [comment]
Weird.
Can you try to separate the two tables with a line (Ctrl+r)

**Comment by Salvaju29ro** [comment]
I tried:

![image](https://user-images.githubusercontent.com/10899871/94990809-14a0cc80-057f-11eb-96fa-d1306eb579e5.png)

However, I made a mistake
The tables overlap when the app opens on another node and then I move to the one with the tables
If Cherrytree opens with the node with tables, they are normal

**Comment by ciranus** [comment]
So let's see if some-else can reproduce (I really cannot), then maybe specific to Arch ?
Just to see: which **libgtk-3-0** version ?

**Comment by Salvaju29ro** [comment]
Do you believe me when I tell you that I probably haven't installed it? (The red ones are installed, the blue ones aren't)

![image](https://user-images.githubusercontent.com/10899871/94992092-cf34cd00-0587-11eb-8426-6ce2bc4c0b5b.png)
------------------------------------------------------------------------------------------------------------------------------------------------------------------
![image](https://user-images.githubusercontent.com/10899871/94992100-dfe54300-0587-11eb-9dfa-f546098be85b.png)

If so, which one should I install?
Maybe that's the problem

**Comment by ciranus** [comment]
**lib32-gtk3** which is installed is probably same as libgtk-3-0.
gtk3 package of Arch is maybe just a meta-package.
I don't know Arch, then cannot help here.

**Comment by Salvaju29ro** [comment]
> **lib32-gtk3** which is installed is probably same as libgtk-3-0.
> gtk3 package of Arch is maybe just a meta-package.
> I don't know Arch, then cannot help here.

I honestly don't know what else to say.
For me it is a negligible "problem", since the notes and tables are not permanently lost.
At this point I'd wait for TXE, who said it happened to him too (and coincidentally he's on Arch too).
@Ciranus Thank you so much for your help, you have been very kind

**Comment by giuspen** [comment]
I have to say I saw such weirdness in the past on my Ubuntu 20.04 when closing the table edit dialog by pressing the return key (which is clicking the button OK) but it went away when I moved the focus to the button before clicking it (button.grab_focus()), I haven't seen it since. This is just to help @txe in case he can see it again in Arch Linux.

**Comment by ciranus** [comment]
Maybe a side effect of Archlinux which is a "_Cutting-Edge_" Distribution....
Cannot be as robust as Debian, or even ubuntu or derivates.

[gtk3 in Arch](https://aur.archlinux.org/packages/?O=0&SeB=b&K=gtk3&outdated=&SB=n&SO=a&PP=50&do_Search=Go)... Good luck to identify the good gtk3 package.

> gtk3-adwaita-3-32-git
> gtk3-classic
> gtk3-cosy
> gtk3-donnatella
> gtk3-git
> gtk3-libwinmenu
> gtk3-mushrooms
> gtk3-nocsd-git
> gtk3-patched-filechooser-icon-view
> gtk3-typeahead
> gtk3-ubuntu

**Comment by txe** [comment]
@giuspen , the issue is really stable on Arch, it looks like drawing glitch because when tables are placed wrong
-  you can start selecting text in a node and see the selected text are misplaced too. It starts redrawing and changing.
- if you move focus from the app, the tables instantly jump into right places. (looks similar to the case with button focus you mention before)

I tried to remove all code or add more redrawing code but there is no difference at all. @giuspen , can you install arch and help me with that, I don't see what I can do else.

**Comment by ciranus** [comment]
[Arch Maintainer: PrinceMachiavell (morgenstern)](https://aur.archlinux.org/packages/cherrytree/) should maybe introduced in the loop of investigation to clarify if it is a CT issue, or an Archlinux gtk3 lib issue already identified on other applications.

**Comment by giuspen** [comment]
@txe I'll install Arch on a VM and have a go

**Comment by ciranus** [comment]
For info, there is also [ArchBang live Iso](https://archbang.org/) which can potentially be helpful for Arch  tests/investigations.

**Comment by crogonint** [comment]
This is probably NOT related, but many many years ago, inserting tables in to MS Word and MS Works documents had the same type of issue. The problem turned out to be an overrun issue in the drawing routine. If I recall, There was some math issue where zooming it at precisely the wrong zoom level. or trying to draw it at precisely the wrong number of pixels mucked up the entire table. Try to do things like make the window smaller, zoom in / out, set the table size to a smaller number of pixels, maybe even shrink down the column sizes. Try each of these repetedly and see if the disapearing table issue stops occurring at some point.

Again, I kind of doubt if this is the exact same issue, but you never know. Good luck!

**Comment by giuspen** [comment]
I experienced a similar glitch with an automatically expanding codebox wider than the cherrytree window, so with an horizontal scrollbar at the bottom of the rich text content. When moving the cursor below the codebox weirdnesses are displayed like duplicated lines of text, but it's just a glitch in the visualisation

**Comment by giuspen** [comment]
should be fixed in upcoming 0.99.17
